### PR TITLE
Add substitution when loading yaml files

### PR DESCRIPTION
**Pull request by nano-meter**
implements #1051

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Can you please add a test for this new functionality.

**Comment by nano-meter** [comment]
Should I instead of creating a new test yaml file just modify params.yaml?

**Comment by alextoind** [comment]
I was exactly searching for this feature right now!

@dirk-thomas do you think you'll backport it also to kinetic-devel when you are happy with the implementation?

File: tools/roslaunch/src/roslaunch/xmlloader.py [hunk 1]
```diff
@@ -235,9 +235,10 @@ def _rosparam_tag(self, tag, context, ros_config, verbose=True):
             # load is the default command            
             cmd = cmd or 'load'
             value = _get_text(tag)
+            subst_function = None
             if subst_value:
-                value = self.resolve_args(value, context)
-            self.load_rosparam(context, ros_config, cmd, param, file, value, verbose=verbose)
+                subst_function = lambda x: self.resolve_args(x, context)
+            self.load_rosparam(context, ros_config, cmd, param, file, value, verbose=verbose, subst_function=subst_function)
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/xmlloader.py:241]: Without the patch even a plain `value` (no `file`) was being substituted when `subst_value` was set. After this change substitution seems to only happen on the file content. Is that intentional? This seems to change the existing behavior which would be undesired.

**Comment by dirk-thomas** [comment]
> do you think you'll backport it also to kinetic-devel when you are happy with the implementation?

That depends if the final patch changes any API or behavior and how big the changes / chances for regressions are.

File: tools/roslaunch/src/roslaunch/xmlloader.py (hunk 1 above)
Comment (nano-meter) [reply tools/roslaunch/src/roslaunch/xmlloader.py:241]: No, I am not changing existing behavior, the code that does the change is moved to 
https://github.com/nano-meter/ros_comm/blob/a7ed7d8c003f52757b75121ee021b578b4b33a07/tools/roslaunch/src/roslaunch/loader.py#L405

This way both value and file content substitution is handled the same way. 
I also created tests for file and value substitution, so if I introduced undesired changes, those probably would have caught it. If you see a problem, please tell me so I could fix it.

Comment (dirk-thomas) [reply tools/roslaunch/src/roslaunch/xmlloader.py:241]: Thanks for clarifying. I see now this is only relevant when `cmd` is `load` where it is handles in the function. The other commands don't mind about `value` so it is ok to not substitute it in these cases.

**Comment by dirk-thomas** [comment]
> Should I instead of creating a new test yaml file just modify params.yaml?

The new separate yaml file looks fine to me.

Thanks for the patch. Checking CI results before merging...

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder retest this please

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder retest this please

**Comment by dirk-thomas** [comment]
Since the failing RosAssert unit tests are unrelated (already on the target branch the case and ticketed separately in #1358) I will go ahead and merge this anyway.

**Comment by miguelprada** [comment]
Now that this has been merged, I'd like to bring back the issue raised by @alextoind of backporting to kinetic-devel. Is there any chance of that happening @dirk-thomas?

**Comment by dirk-thomas** [comment]
> Is there any chance of that happening @dirk-thomas?

That will be considered before the next Kinetic release. In general new features have a lower chance of getting backported.

**Comment by tlind** [comment]
Would also be interested in getting this backported to ROS Kinetic, given that it still has a lifetime of 3 years and this feature looks quite elementary.

**Pull request merged**